{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24d199bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Maybe a good next step would be to create a phylogenetic tree of tRNA sequences, and try to see \n",
    "if certain isodecoders cluster together, rather than by phylum.\n",
    "\n",
    "Here, I will write a script to pull out (aligned) specific isotype sequences and output them to fasta format.\n",
    "\n",
    "Next, I will download those sequences and build a phylogenetic tree in jalview. I will use that tree to see\n",
    "if specific isodecoders cluster together--which would represent functionally similar isodecoders\n",
    "\n",
    "Using my GtRNAdb-scanning framework\n",
    "\"\"\"\n",
    "\n",
    "import sys\n",
    "sys.path.append('..')\n",
    "sys.path.append('../sprinzl_gtrnadb/')\n",
    "sys.path.append('../cca_tails/')\n",
    "import pandas as pd\n",
    "from nameSprinzlPositions import sprinzlAlign\n",
    "import os\n",
    "\n",
    "def parseGtRNAdbGenomes(inFile):\n",
    "    \"\"\"parse GtRNAdb genomes files\"\"\"\n",
    "    info = {}\n",
    "    chrSizes = {}\n",
    "    chrKey = {}\n",
    "    genKey = {}\n",
    "    abbrKey = {}\n",
    "    phylKey = {}\n",
    "    spKey = {}\n",
    "    with open(inFile, 'r') as inF:\n",
    "        entries = inF.read().split('\\n\\n')\n",
    "        \n",
    "        for entry in entries[1:-1]:\n",
    "            lines = entry.split('\\n')\n",
    "            \n",
    "            abbr = lines[0].split(' ')[-1]\n",
    "            acc = lines[7].split(' ')[-1]\n",
    "            db = lines[4].split(' ')[-1]\n",
    "            chrSizes[abbr] = {}\n",
    "            chrKey[abbr] = {}\n",
    "            genKey[lines[7].split(' ')[-1]] = abbr\n",
    "            abbrKey[abbr] = db\n",
    "            \n",
    "            phylKey[db] = ' '.join(lines[11].split(' ')[5:-1])\n",
    "            \n",
    "            sp = ' '.join(lines[6].split(' ')[3:])\n",
    "            #determine species name, handle special cases\n",
    "            if sp.split(' ')[1] == 'sp.':\n",
    "                spName = sp\n",
    "\n",
    "            elif sp[0] == sp[0].upper():\n",
    "                spName = ' '.join(sp.split(' ')[0:2])\n",
    "\n",
    "            else:\n",
    "                spName = sp\n",
    "                \n",
    "            spKey[db] = spName\n",
    "                \n",
    "            \n",
    "            for x, y, z in zip(lines[19].split(' ')[3:], lines[21].split(' ')[3:], lines[20].split(' ')[5:]):\n",
    "                if len(x.split('_')) > 1:\n",
    "                    chrSizes[abbr]['_'.join(x.split('_')[1:])] = y\n",
    "                else:\n",
    "                    chrSizes[abbr][x] = y\n",
    "                    \n",
    "                chrKey[abbr][z] = x\n",
    "            \n",
    "            info[abbr] = acc\n",
    "        inF.close()\n",
    "            \n",
    "    return info, chrSizes, chrKey, genKey, abbrKey, phylKey, spKey\n",
    "\n",
    "def formatNameMap(f):\n",
    "    data = pd.read_csv(f)\n",
    "    \n",
    "    #print(data['GTRNADB_ID'])\n",
    "    data['LEGACY_ID'] = [x.replace('trna', 'tRNA') for x in data['LEGACY_ID']]\n",
    "    \n",
    "    nm = data[['GTRNADB_ID']].set_index(data['LEGACY_ID'] )\n",
    "    \n",
    "    return nm\n",
    "\n",
    "\n",
    "def main():\n",
    "    # The goal of this script is to find all the unique tRNA sequences in GtRNAdb, and\n",
    "    # organize them by isotype, while retrieving alignments of the tRNA sequence; the\n",
    "    # number of occurences of the sequence, and the assembly they occur in.\n",
    "    \n",
    "    acLookup = {'Ala': ['AGC', 'GGC', 'TGC', 'CGC'], 'Gly': ['ACC', 'GCC', 'TCC', 'CCC'], \n",
    "                'Pro': ['AGG', 'GGG', 'CGG', 'TGG'], 'Thr': ['AGT', 'GGT', 'CGT', 'TGT'],\n",
    "                'Val': ['AAC', 'GAC', 'TAC', 'CAC'],\n",
    "                'Ser': ['AGA', 'GGA', 'CGA', 'TGA', 'ACT', 'GCT'],\n",
    "                'Arg': ['ACG', 'GCG', 'TCG', 'CCG', 'TCT', 'CCT'],\n",
    "                'Leu': ['AAG', 'TAG', 'GAG', 'CAG', 'CAA', 'TAA'],\n",
    "                'Phe': ['AAA', 'GAA'], 'Asn': ['ATT', 'GTT'],\n",
    "                'Lys': ['CTT', 'TTT'], 'Asp': ['ATC', 'GTC'],\n",
    "                'Glu': ['CTC', 'TTC'], 'His': ['ATG', 'GTG'],\n",
    "                'Gln': ['CTG', 'TTG'], 'Ile': ['AAT', 'GAT', 'TAT'],\n",
    "                'Ile2': ['CAT'], 'Met': ['CAT'], 'iMet': ['CAT'], 'fMet': ['CAT'],\n",
    "                'Tyr': ['GTA', 'ATA'], 'Sup': ['TCA', 'CTA', 'TTA'],\n",
    "                'Cys': ['ACA', 'GCA'], 'Trp': ['CCA'], 'SeC': ['TCA', 'CTA', 'AAA'], 'Und': ['NNN'] }\n",
    "    \n",
    "    domain = 'Bacteria'\n",
    "    \n",
    "    info, chrSizes, chrKey, genKey, abbrKey, phylKey, spKey = parseGtRNAdbGenomes('../gtrnadb_genome_info.txt')\n",
    "    \n",
    "    #Stores dictionary of pandas DataFrame objects which store isodecoder sequence, and associated assembly as axes\n",
    "    isoacceptorCounts = {}\n",
    "    for x in acLookup.keys():\n",
    "        for y in acLookup[x]:\n",
    "            isoacceptorCounts[x + '-' + y] = []\n",
    "            \n",
    "    isoacceptorCounts = pd.DataFrame(isoacceptorCounts)\n",
    "    isoacceptorCounts = isoacceptorCounts[sorted(isoacceptorCounts.columns)] #Reorder columns alphabetically\n",
    "    \n",
    "    #Iterate through species in GtRNAdb\n",
    "    for abbr in list(info.keys()):#[-175:-174]: #<--- specifically test vibrChol1 or whatever\n",
    "        #keys for relevant info\n",
    "        db = abbrKey[abbr]\n",
    "        ph = phylKey[db]\n",
    "        sp = spKey[db]\n",
    "        print(db)\n",
    "        isoacceptorCounts.loc[db, :] = 0\n",
    "        \n",
    "        namemap = pd.DataFrame()\n",
    "        \n",
    "        #retrieve relevant sequence information\n",
    "        try:\n",
    "            namemap = formatNameMap('/projects/lowelab/users/pchan/GtRNAdb2/tRNA-runs/{2}/{0}/{1}_export.csv'.format(abbr, db, domain))\n",
    "            \n",
    "\n",
    "        #Handle inconsistent naming problems (like vibrChol1)\n",
    "        except FileNotFoundError:\n",
    "\n",
    "            try:\n",
    "\n",
    "                namemap = pd.DataFrame()\n",
    "\n",
    "                for x in os.listdir('/projects/lowelab/users/pchan/GtRNAdb2/tRNA-runs/{1}/{0}/'.format(abbr, domain)):\n",
    "\n",
    "                    if x.split('_')[-1] == 'export.csv':\n",
    "                        namemap = formatNameMap('/projects/lowelab/users/pchan/GtRNAdb2/tRNA-runs/{2}/{0}/{1}'.format(abbr, x, domain))\n",
    "                    else:\n",
    "                        pass\n",
    "            #Warn about issues beyond inconsistent \n",
    "            except FileNotFoundError:\n",
    "                print('warning: could not access files for ' + db)\n",
    "        \n",
    "        \n",
    "        for x in namemap.index.values:\n",
    "            name = namemap.loc[x, 'GTRNADB_ID']\n",
    "            \n",
    "            acc = '-'.join(name.split('-')[1:3])\n",
    "            \n",
    "            try:\n",
    "                isoacceptorCounts.loc[db, acc] += 1\n",
    "            except KeyError:\n",
    "                isoacceptorCounts[acc] = 0\n",
    "                isoacceptorCounts.loc[db, acc] += 1\n",
    "            \n",
    "    isoacceptorCounts.to_csv('./gtrnadb_isoacceptor_counts.txt', sep = '\\t')\n",
    "        \n",
    "        \n",
    "    \n",
    "def writeAlignmentColumns(d, file):\n",
    "    with open(file, 'w') as f:\n",
    "        \n",
    "        f.write(','.join(d))\n",
    "        f.close()\n",
    "        \n",
    "    \n",
    "if __name__ == '__main__':\n",
    "    main()\n",
    "    \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
